File: Matlab/SCRATCH.m
% SCRATCH.

%% Load data.
clc
clear
% load('OOdata (normalized).mat')
%load('Coffee-normalized.mat')
 load('ECGdata (normalized).mat')
[n,p] = size(train);
p = p-1;


%% Split testing data as validation and testing set.
valratio = 0.5; % Extract this fraction for validation set.
[val, test] = train_test_split(test, valratio);



%%
%prepare the data set
gmults = linspace(0, 1.5, 15);
beta=2;
tol.rel = 1e-3;
tol.abs= 1e-3;
maxits= 1000;
quiet=false;

consttype = 'sphere';

sparsity_level = 0.5;

D = eye(p);

%% Call validation set.
[val_w, DVs, gamma, best_ind, val_score, classMeans] = PenZDAval(train, val,D, gmults, consttype, sparsity_level, beta, tol, maxits,quiet);
gamma, best_ind

stats = predict(val_w, test, classMeans)

% Plot DVs.
[~,K] = size(classMeans);
for i = 1:K-1
    figure
    plot(1:p, val_w(:,i))
end

%% Try cross-validation.

nfolds = n;
[bestDVs, bestind, bestgamma,  cv_scores, classMeans] = PenZDAcv(train, nfolds, D, gmults, consttype, sparsity_level, beta, tol, maxits,quiet);

bestind
bestgamma
cvstats = predict(bestDVs, test, classMeans)

% Plot DVs.
[~,K] = size(classMeans);
for i = 1:K-1
    figure
    plot(1:p, bestDVs(:,i))
end




%% Call solver.
gamscale = 0.3;
tic;
pentype = 'ball';
[DVs,~,~,~,classMeans, ~] = PenZDA(train,D,tol,maxits,beta,quiet, pentype,gamscale);
             
t0 = toc, % Stop timer after training is finished.
        
stats0 = test_ZVD_V1(DVs,test,classMeans)

% [~,K] = size(classMeans);
% for i = 1:K-1
%     figure
%     plot(DVs(:,i))
% end
% 

% Call spherical solver.
tic
pentype = 'sphere';
[DVs1,~,~,~,classMeans,gamma] = PenZDA(train,D, tol,maxits,beta,quiet, pentype,gamscale);
t1 = toc, % Stop timer after training is finished.
        
stats1 = test_ZVD_V1(DVs1,test,classMeans)

% Plot DVs.
[~,K] = size(classMeans);
for i = 1:K-1
    figure
    plot(1:p,DVs1(:,i),  1:p,DVs(:,i) )
end


%% Check accuracy.

t0 - t1
norm(DVs - DVs1)

File: R/SCRATCH.R
# Make ASDA Input.
n <- dim(train)[1]
p <- dim(train)[2] - 1

Xtrain <- as.matrix(train[, 2:(p+1)])
Ytrain <- as.factor(train$X1)

Xtest <- as.matrix(test[, 2:(p+1)])
Ytest <- as.factor(test$X1)

ntest <- dim(test)[1]



#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# Call ASDA.
library(accSDA)

res <- ASDA(Xt = Xtrain, Yt = Ytrain)

# Predict on the test data
preds <- predict(res, newdata = Xtest)

# Calculate accuracy
sum(preds$class == Ytest)/ntest # We have N samples per class, so total 3*N

# Plot DV.
plot(res$beta, type = 'l')


#++++++++++++++++++++++++++++++++++++++++++++++++++++
# Test penzda.
library(MASS)
library(rARPACK)

cmns <- penzda(Xt = Xtrain, Yt = Ytrain, maxits=50, tol = 1e-3, type ="ball")
cmns$DVs
plot(cmns$DVs[,1], type="l")


#++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# TEST VALIDATION SCHEME.
#++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

# Split data.
set.seed(2)
nval <- round(0.2*n)
valinds <-  sample.int(n, size = nval, replace = FALSE)

Xval <- Xtrain[valinds, ]
Yval <- Ytrain[valinds]
Xvt <- Xtrain[-valinds, ]
Yvt <- Ytrain[-valinds]

# Need to renormalize training and validation data.
trainlist <- normalize(x = Xvt)
Xvt <- trainlist$x
mu <- trainlist$mu
sig <- trainlist$sig

Xval <- normalizetest(x=Xval, mu = mu, sig = sig)

res <- penzdaVAL(Xt = Xvt, Yt = Yvt, Xval = Xval, Yval=Yval, maxits = 500,
                 gmults = c(0.25, 0.5, 0.75, 1),sparsity_level = 0.4,
                 quiet = FALSE,type = "sphere")
                    

plot(res$DVs, type="l")
res$val_score

penstats <- predict(obj = res, Xtest = Xtest, Ytest = Ytest)
penstats$mc

# TEST VALIDATION SCHEME.
#++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

cvres <- penzdaCV(Xt = Xtrain, Yt = Ytrain, nfolds = 5, maxits = 500,
                 gmults = c(0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2,5),sparsity_level = 0.4,
                 quiet = FALSE,type = "sphere")

plot(cvres$DVs, type="l")
cvres$cvscores

rowMeans(cvres$cvscores)

penstats <- predict(obj = cvres, Xtest = Xtest, Ytest = Ytest)

penstats$mc
penstats$l0/136

cbind(penstats$preds, Ytest, penstats$dist)

# v  <- c(1, 2, -4)
# a <- 1.5
# 
# s <- vecshrink(v,a)

#++++++++++++++++++++++++++++++++
x <- cmns$sols$x
y <- cmns$sols$y
z <- cmns$sols$z
gamma <- cmns$gamma
D <- diag(p)



# r <- rbind(c(1,2,3),
#            c(0,2,1),
#            c(0,0,2))
# 
# prds <- max.col(r)
# r

# y <- backsolve(r, x <- c(8,4,2)) 
# 
# r %*% y
# 
# backsolve(r, x, transpose = TRUE)

